pkg/internal/apis/certmanager/validation/util/nameserver.go

```diff
@@ -24,43 +24,47 @@ import (
 
 var defaultRFC2136Port = "53"
 
-// This function make a valid nameserver as per RFC2136
+// This function returns a valid nameserver (in the form <host>:<port>) for the RFC2136 provider
 func ValidNameserver(nameserver string) (string, error) {
+	nameserver = strings.TrimSpace(nameserver)
 
 	if nameserver == "" {
 		return "", fmt.Errorf("RFC2136 nameserver missing")
 	}
 
+	// When passed an IPv6, SplitHostPort expect it to be enclosed in square brackets
+	if ipaddr := net.ParseIP(nameserver); ipaddr != nil && ipaddr.To4() == nil {
+		nameserver = "[" + ipaddr.String() + "]:" + defaultRFC2136Port
+	}
```

Comment:
Instead of having this logic, would it make sense to require *users* to encompass the specified IPv6 address within `[` and `]`? Why the magic handling here?

Reply:
The whole issue with enclosed IPv6 here is that `net.SplitHostPort` expect the IPv6 to be enclosed in square brackets (which I can understand in its context) but on the other hand, `net.JoinHostPort` don’t work correctly with enclosed IPv6: if it thinks the host is an IPv6 address, it will enclose it in square brackets no matter what (e.g. `net.JoinHostPort("[2001:db8::1]", "53")` will return `[[2001:db8::1]]:53`).

I find it really strange to have to enclose an IPv6 address when no ports are specified. Also, if the user set a non-enclosed IPv6, the only way to report an error that’s usefull will be to do the same validation : if `nameserver` is a valid (non enclosed) IPv6 address, then raise an error.

Reply:
Looking here: https://en.wikipedia.org/wiki/IPv6_address#Literal_IPv6_addresses_in_network_resource_identifiers

It seems like the reason for using square brackets is to disambiguate between port numbers and IPv6 addresses that end with numerical 'segments'.

Given that, I think it makes sense that we require IPv6 addresses to be specified within square brackets from a user perspective, even if the user does not explicitly specify a port number.

Reply:
So just to be sure, is this ok for you (the first one is most the common way of writing IPv6 address):

```
Input             Current state (PR)  Proposed state
2001:db8::1       Valid               Invalid (report error?)
[2001:db8::1]     Valid               Valid
```

Should we check if `nameserver` is set to an unenclosed IPv6 address (first item) and raise an error?

Reply:
I came with another solution that you might prefer, it allows IPv6 addresses to be set either enclosed or unenclosed, but removes that logic of checking for an IPv6 address and enclosing it: https://play.golang.org/p/XQlDwwNKaO-

Reply:
I'd prefer to keep the number of ways a user *can* specify this address down to just one, which should then be validated by our validation webhook. I'm not sure if it's really desirable or useful to support unenclosed IPv6 addresses given it is recommended to enclose addresses, especially if they include port numbers.

For example, if we don't enclose the address, this string becomes ambiguous/non-obvious to the reader: `2001:db8:85a3:8d3:1319:8a2e:370:5553`

Reply:
> I'd prefer to keep the number of ways a user can specify this address down to just one

Duly noted, I’ll update the PR.

> For example, if we don't enclose the address, this string becomes ambiguous/non-obvious to the reader: `2001:db8:85a3:8d3:1319:8a2e:370:5553`

I don’t want to keep this debate going forever, but I think it’s well known that an IPv6 address must be enclosed when followed by a port number. Also, the example you gave is not really misleading as `2001:db8:85a3:8d3:1319:8a2e:370` (the same address without what could be the port number) is not a valid IPv6 address. A confusing setting could be `2001:db8::370:5553` vs `2001:db8::370` (both are valid).

Anyway, it’s you choice, I’m not arguing.

Reply:
Ok, done.

I updated the documentation for the `nameserver` field and the error message in `issuer.go` to reflect that IPv6 address must always be enclosed and I also kept the test case for unenclosed IPv6 address.